Add README for comms (#46)
leancodepl/comms: packages/comms/README.md

you'd have to use cables, but in code we can use `comms` for them to communicate
easily without having to "tear down the walls".

### Creating a Listener

```dart
/// Add a `Listener` mixin with type of messages to listen for.
class LightBulb with Listener<bool> {
  LightBulb() {
    /// Call `listen` to start listening
    listen();
  }

  bool enabled = false;
  
  /// Override `onMessage` to specify how to react on messages.
  @override
  void onMessage(bool message) {
    enabled = message;
  }

  void dispose() {
    /// Call `cancel` to stop listening and clean up.
    cancel();
  }
}
```

### Creating a Sender

```dart
/// Add a `Sender` mixin with type of messages to send.
class LightSwitch with Sender<bool> {
  void enable() {
    /// Call `send` to deliver the message to a Listener.
    send(true);
  }
}
```

### Using Listener and Sender

```dart
void main() async {
  // Just create instances of both classes, comms will 
  // handle connection between them.
  final lightBulb = LightBulb();
  final lightSwitch = LightSwitch();

  print(lightBulb.enabled); // false

  lightSwitch.enable();
    
  // Because communication is asynchronuos we have to wait

>   // just a little bit for the message to reach the `lightBulb`.

Author: well yeah, but I don't want to go into details of asynchrony in dart here, and just explain why if anyone copied this code, it wouldn't work without this line, if you have any ideas I am open to changing it

  await Future<void>.delayed(Duration.zero);

  print(lightBulb.enabled); // true

  // At the end don't forget to close the connection and cleanup.
  lightBulb.dispose();
}
```

### Multiple Listeners and Senders

> With `comms` any `Sender` will send a message to any `Listener` sharing the same
> message type.

Reviewer: ```suggestion

A `Sender` sends a message to all `Listener` sharing the same
message type.
```

[See this](https://en.wikiversity.org/wiki/Technical_writing/Style#Understanding_present_tense)

Also, I'd add one more sentence with some example:

```
For example, if you have a single `Sender<UserData>`, and 3 `Listener<UserData`, then whenever that sender sends a message, all the listeners will receive the message (simultaneously? idk).
```

```dart
void main() async {
  final lightBulbA = LightBulb();
  final lightSwitchA = LightSwitch();
  
  final lightBulbB = LightBulb();
  final lightSwitchB = LightSwitch();

  print(lightBulbA.enabled); // false
  print(lightBulbB.enabled); // false

  lightSwitchB.enable();

  print(lightBulbA.enabled); // true
  print(lightBulbB.enabled); // true
}
```

## Handling initial message
To handle the last message sent before creating an instance of `Listener` you

> can use `onInitialMessage` which will be invoked after `Listener` calls `listen`

Author: That's ok but personally I don't like adding ()

  void onMessage(CounterMessage message) {
    if (message is CounterIncremented) {
      count--;
    }
    if (message is CounterDecremented) {
      count++
    }
  }

  @override
  void onInitialMessage(CounterMessage message) => onMessage(message);
}

void main() {
  final counterController = CounterController();

  counterController.increment();

  final counter = Counter();
    
  print(counter.count); // 1
}
```

## Listening for multiple message types
If you need to receive more than one message type in a single `Listener` you
can use `MultiListener` mixin.

```dart
class MyListener with MultiListener {
  MyListener() {
    listen();
  }

  @override
  List<ListenerDelegate> get listenerDelegates => [
        ListenerDelegate<CounterMessage>(),
        ListenerDelegate<AuthMessage>(),
      ];

  @override
  void onMessage(dynamic message) {
    if (message is CounterMessage) {...}
    if (message is AuthMessage) {...}
  }
}
```

## Custom Senders
To create a custom `Sender` for example to send multiple message types, you
 can use `getSend` which returns `send` method for messages of passed type.

```dart
mixin CustomSender {
  final sendString = getSend<String>();

>   final sendInt = getSend<int>();

Reviewer: On the other hand, this is a mini tutorial, so maybe simple types make more sense.

I'm torn.

>   final sendInt = getSend<int>();

Author: I don't want to advise anyone on how to use this package, also I think that `Multiple Listeners and Senders` section kind of means that

>   final sendInt = getSend<int>();

Author: Also in this example I wanted to show that you can easily use both functions without showing custom types definitions

  void sendMessages() {
    sendString('hello');
    sendInt(1);
  }
}
```

> You can also use `getSend` to send messages from anywhere without even

Reviewer: ```suggestion
You can also use the `getSend()` function to send messages from anywhere without even
```

# comms

> [![codestyle][pub_badge_style]][pub_badge_link]

Reviewer: add badge with a link to pub.dev

[example here](https://raw.githubusercontent.com/leancodepl/patrol/master/README.md)

Author: I'll add it to the main readme later

`comms` is a simple communication pattern abstraction on streams, created for
communication between any classes. It allows `Listener`s to easily react to
messages sent by `Sender`s.

> For use in Flutter projects, check out [flutter_comms].

Reviewer: It'd be useful to explain what benefits `flutter_comms` brings – e.g automatic `dispose()`

Author: I'd expect to see it there in readme, first line or sth

Reviewer: ah right, makes sense

Reviewer: since you're on writing docs frenzy right now, why not write a cool readme for `flutter_comms` now as well? 👅

Author: I need a break 😄
